File: breeze-app/app/src/main/java/com/mtkresearch/breeze_app/utils/ChatMessageAdapter.java
package com.mtkresearch.breeze_app.utils;

import android.net.Uri;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;
import com.mtkresearch.breeze_app.R;

import java.util.ArrayList;
import java.util.List;

/**
 * Adapter for displaying chat messages in a RecyclerView.
 * Handles only UI representation of messages.
 */
public class ChatMessageAdapter extends RecyclerView.Adapter<ChatMessageAdapter.MessageViewHolder> {
    private static final String TAG = "ChatMessageAdapter";
    private final List<ChatMessage> messages = new ArrayList<>();
    private final Handler handler = new Handler(Looper.getMainLooper());
    private OnSpeakerClickListener speakerClickListener;
    private OnMessageLongClickListener messageLongClickListener;
    private OnFeedbackClickListener feedbackClickListener;
    private boolean ttsEnabled = AppConstants.TTS_ENABLED;  // Default to AppConstants value

    public interface OnSpeakerClickListener {
        void onSpeakerClick(String messageText, int position);
    }

    public interface OnMessageLongClickListener {
        boolean onMessageLongClick(ChatMessage message, int position);
    }

    public interface OnFeedbackClickListener {
        void onFeedbackClick(boolean isUpvote);
    }

    public void setSpeakerClickListener(OnSpeakerClickListener listener) {
        this.speakerClickListener = listener;
    }

    public void setOnMessageLongClickListener(OnMessageLongClickListener listener) {
        this.messageLongClickListener = listener;
    }

    public void setFeedbackClickListener(OnFeedbackClickListener listener) {
        this.feedbackClickListener = listener;
    }

    public void setTTSEnabled(boolean enabled) {
        this.ttsEnabled = enabled;
        notifyDataSetChanged();  // Refresh all items to update speaker icon visibility
    }

    public void setMessages(List<ChatMessage> newMessages) {
        messages.clear();
        messages.addAll(newMessages);
        notifyDataSetChanged();
    }

    public void addMessage(ChatMessage message) {
        messages.add(message);
        notifyItemInserted(messages.size() - 1);
    }

    public void updateLastMessage(String newText) {
        if (!messages.isEmpty()) {
            int lastIndex = messages.size() - 1;
            ChatMessage lastMessage = messages.get(lastIndex);
            lastMessage.updateText(newText);
            notifyItemChanged(lastIndex);
        }
    }

    public void removeLastMessage() {
        if (!messages.isEmpty()) {
            int lastIndex = messages.size() - 1;
            messages.remove(lastIndex);
            notifyItemRemoved(lastIndex);
        }
    }

    public void clearMessages() {
        messages.clear();
        notifyDataSetChanged();
    }

    public List<ChatMessage> getMessages() {
        return messages;
    }

    public void setMessageTextColor(int position, int color) {
        if (position >= 0 && position < messages.size()) {
            messages.get(position).setCustomTextColor(color);
            notifyItemChanged(position);
        }
    }

    @Override
    public int getItemViewType(int position) {
        return messages.get(position).isUser() ? 0 : 1; // 0 for user, 1 for AI
    }

    @NonNull
    @Override
    public MessageViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view;
        if (viewType == 0) {
            // User message layout
            view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.item_chat_message_user, parent, false);
            return new UserMessageViewHolder(view);
        } else {
            // AI message layout
            view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.item_chat_message_ai, parent, false);
            return new AIMessageViewHolder(view);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull MessageViewHolder holder, int position) {
        ChatMessage message = messages.get(position);
        
        // Set text selection mode
        holder.messageText.setTextIsSelectable(true);
        
        // Set long click listener
        holder.itemView.setOnLongClickListener(v -> {
            if (messageLongClickListener != null) {
                return messageLongClickListener.onMessageLongClick(message, position);
            }
            return false;
        });

        // Set the message text
        holder.messageText.setText(message.getText());

        // Set the text color based on custom color if set
        if (message.getCustomTextColor() != 0) {
            holder.messageText.setTextColor(message.getCustomTextColor());
        } else if (message.isUser()) {
            holder.messageText.setTextColor(holder.itemView.getContext().getColor(R.color.user_message_text));
        } else {
            holder.messageText.setTextColor(holder.itemView.getContext().getColor(R.color.ai_message_text));
        }

        // Get the ConstraintLayout params for the message bubble
        ConstraintLayout.LayoutParams params = (ConstraintLayout.LayoutParams) holder.messageBubble.getLayoutParams();

        if (message.isUser()) {
            setupUserMessage(holder, params, message);
        } else {
            setupAssistantMessage(holder, params, message);
        }

        holder.messageBubble.setLayoutParams(params);
        setupImageAndSpeakerButtons(holder, message, position);
    }

    private void setupUserMessage(MessageViewHolder holder, ConstraintLayout.LayoutParams params, ChatMessage message) {
        if (holder instanceof UserMessageViewHolder) {
            UserMessageViewHolder userHolder = (UserMessageViewHolder) holder;
            holder.messageBubble.setBackgroundResource(R.drawable.bg_user_message);
            
            // Show button row if message has text and TTS is enabled
            if (message.hasText() && ttsEnabled) {
                userHolder.buttonRow.setVisibility(View.VISIBLE);
                userHolder.speakerButton.setVisibility(View.VISIBLE);
            } else {
                userHolder.buttonRow.setVisibility(View.GONE);
            }

            // Keep upper text row hidden for now (future extension)
            userHolder.upperTextRow.setVisibility(View.GONE);
        }
    }

    private void setupAssistantMessage(MessageViewHolder holder, ConstraintLayout.LayoutParams params, ChatMessage message) {
        if (holder instanceof AIMessageViewHolder) {
            AIMessageViewHolder aiHolder = (AIMessageViewHolder) holder;
            holder.messageBubble.setBackgroundResource(R.drawable.bg_ai_message);
            
            // Show/hide button row based on message completion and TTS state
            if (message.isCompleted()) {
                aiHolder.buttonRow.setVisibility(View.VISIBLE);
                
                // Show/hide speaker button based on TTS state and if not an error message
                boolean showSpeaker = ttsEnabled && message.hasText() && !message.isError();
                aiHolder.speakerButton.setVisibility(showSpeaker ? View.VISIBLE : View.GONE);
                
                // Setup up/down like buttons
                setupFeedbackButtons(aiHolder);
            } else {
                aiHolder.buttonRow.setVisibility(View.GONE);
            }
        }
    }

    private void setupFeedbackButtons(AIMessageViewHolder holder) {
        // Show feedback buttons (always visible when message is complete)
        holder.upLikeButton.setVisibility(View.VISIBLE);
        holder.downLikeButton.setVisibility(View.VISIBLE);
        
        // Set click listeners for feedback buttons
        holder.upLikeButton.setOnClickListener(v -> {
            if (feedbackClickListener != null) {
                feedbackClickListener.onFeedbackClick(true); // true for upvote
            }
        });
        
        holder.downLikeButton.setOnClickListener(v -> {
            if (feedbackClickListener != null) {
                feedbackClickListener.onFeedbackClick(false); // false for downvote
            }
        });
    }

    private void setupImageAndSpeakerButtons(MessageViewHolder holder, ChatMessage message, int position) {
        Uri imageUri = message.getImageUri();
        if (imageUri != null) {
            Log.d(TAG, "Image URI present: " + imageUri);
            holder.messageImage.setVisibility(View.VISIBLE);
            try {
                holder.messageImage.setImageURI(null);
                holder.messageImage.setImageURI(imageUri);
            } catch (Exception e) {
                Log.e(TAG, "Error loading image", e);
                holder.messageImage.setVisibility(View.GONE);
            }
        } else {
            holder.messageImage.setVisibility(View.GONE);
        }

        setupSpeakerClickListeners(holder, message, position);
    }

    private void setupSpeakerClickListeners(MessageViewHolder holder, ChatMessage message, int position) {
        View.OnClickListener speakerListener = v -> {
            if (speakerClickListener != null && message.hasText()) {
                speakerClickListener.onSpeakerClick(message.getText(), position);
            }
        };

        if (holder instanceof UserMessageViewHolder) {
            UserMessageViewHolder userHolder = (UserMessageViewHolder) holder;
            userHolder.speakerButton.setOnClickListener(speakerListener);
        } else if (holder instanceof AIMessageViewHolder) {
            AIMessageViewHolder aiHolder = (AIMessageViewHolder) holder;
            aiHolder.speakerButton.setOnClickListener(speakerListener);
        }
    }

    @Override
    public int getItemCount() {
        return messages.size();
    }

    static abstract class MessageViewHolder extends RecyclerView.ViewHolder {
        protected final TextView messageText;
        protected final LinearLayout messageBubble;
        protected final ImageView messageImage;

        MessageViewHolder(@NonNull View itemView) {
            super(itemView);
            messageText = itemView.findViewById(R.id.messageText);
            messageBubble = itemView.findViewById(R.id.messageBubble);
            messageImage = itemView.findViewById(R.id.messageImage);
        }
    }

    static class UserMessageViewHolder extends MessageViewHolder {
        private final LinearLayout upperTextRow;
        private final LinearLayout buttonRow;
        private final TextView upperText;
        private final ImageButton speakerButton;

        UserMessageViewHolder(@NonNull View itemView) {
            super(itemView);
            upperTextRow = itemView.findViewById(R.id.upperTextRow);
            buttonRow = itemView.findViewById(R.id.buttonRow);
            upperText = itemView.findViewById(R.id.upperText);
            speakerButton = itemView.findViewById(R.id.speakerButton);
        }
    }

    static class AIMessageViewHolder extends MessageViewHolder {
        private final LinearLayout buttonRow;
        private final ImageButton speakerButton;
        private final ImageButton upLikeButton;
        private final ImageButton downLikeButton;

        AIMessageViewHolder(@NonNull View itemView) {
            super(itemView);
            buttonRow = itemView.findViewById(R.id.buttonRow);
            speakerButton = itemView.findViewById(R.id.speakerButton);
            upLikeButton = itemView.findViewById(R.id.upLikeButton);
            downLikeButton = itemView.findViewById(R.id.downLikeButton);
        }
    }
}

File: Android/app/src/main/java/com/mtkresearch/breezeapp/ui/chat/ChatMessageAdapter.kt
package com.mtkresearch.breezeapp.ui.chat

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.mtkresearch.breezeapp.R
import com.mtkresearch.breezeapp.data.models.ChatMessage
import com.mtkresearch.breezeapp.data.models.MediaType
import com.mtkresearch.breezeapp.data.models.MessageSender
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Adapter for displaying chat messages in a RecyclerView
 */
class ChatMessageAdapter(
    private val onMediaClicked: ((ChatMessage) -> Unit)? = null,
    private val onTtsRequested: ((ChatMessage) -> Unit)? = null
) : ListAdapter<ChatMessage, RecyclerView.ViewHolder>(ChatMessageDiffCallback()) {

    companion object {
        private const val VIEW_TYPE_USER = 0
        private const val VIEW_TYPE_ASSISTANT = 1
        private const val VIEW_TYPE_SYSTEM = 2
        private const val VIEW_TYPE_MEDIA = 3
        private val TIME_FORMATTER = SimpleDateFormat("h:mm a", Locale.getDefault())
    }

    override fun getItemViewType(position: Int): Int {
        val message = getItem(position)
        return when {
            // If it's a user message with media but no text content, show it as a media message
            message.sender == MessageSender.USER && 
                message.mediaUri != null && 
                message.content.isBlank() -> VIEW_TYPE_MEDIA
            message.sender == MessageSender.USER -> VIEW_TYPE_USER
            message.sender == MessageSender.ASSISTANT -> VIEW_TYPE_ASSISTANT
            else -> VIEW_TYPE_SYSTEM
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            VIEW_TYPE_USER -> {
                UserMessageViewHolder(
                    LayoutInflater.from(parent.context)
                        .inflate(R.layout.item_user_message, parent, false),
                    onTtsRequested,
                    onMediaClicked
                )
            }
            VIEW_TYPE_ASSISTANT -> {
                AssistantMessageViewHolder(
                    LayoutInflater.from(parent.context)
                        .inflate(R.layout.item_assistant_message, parent, false),
                    onTtsRequested
                )
            }
            VIEW_TYPE_MEDIA -> {
                MediaMessageViewHolder(
                    LayoutInflater.from(parent.context)
                        .inflate(R.layout.item_media_message, parent, false),
                    onMediaClicked
                )
            }
            else -> {
                SystemMessageViewHolder(
                    LayoutInflater.from(parent.context)
                        .inflate(R.layout.item_system_message, parent, false)
                )
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val message = getItem(position)
        when (holder) {
            is UserMessageViewHolder -> holder.bind(message)
            is AssistantMessageViewHolder -> holder.bind(message)
            is MediaMessageViewHolder -> holder.bind(message)
            is SystemMessageViewHolder -> holder.bind(message)
        }
    }

    /**
     * ViewHolder for media-only messages
     */
    class MediaMessageViewHolder(
        itemView: View,
        private val onMediaClicked: ((ChatMessage) -> Unit)?
    ) : RecyclerView.ViewHolder(itemView) {
        
        private val mediaImageView: ImageView = itemView.findViewById(R.id.mediaImage)
        private val expandButton: ImageButton = itemView.findViewById(R.id.expandButton)
        private var currentMessage: ChatMessage? = null
        
        init {
            mediaImageView.setOnClickListener {
                currentMessage?.let { message ->
                    onMediaClicked?.invoke(message)
                }
            }
            
            expandButton.setOnClickListener {
                currentMessage?.let { message ->
                    onMediaClicked?.invoke(message)
                }
            }
        }
        
        fun bind(message: ChatMessage) {
            currentMessage = message
            
            // Set the media image
            if (message.mediaUri != null && message.mediaType == MediaType.IMAGE) {
                mediaImageView.setImageURI(message.mediaUri)
            }
        }
    }

    /**
     * ViewHolder for user messages
     */
    class UserMessageViewHolder(
        itemView: View,
        private val onTtsRequested: ((ChatMessage) -> Unit)?,
        private val onMediaClicked: ((ChatMessage) -> Unit)? = null
    ) : RecyclerView.ViewHolder(itemView) {
        
        private val messageText: TextView = itemView.findViewById(R.id.userMessageText)
        private val mediaImageView: ImageView? = itemView.findViewById(R.id.userMediaImage)
        private val copyButton: ImageButton = itemView.findViewById(R.id.copyButton)
        private val speakerButton: ImageButton = itemView.findViewById(R.id.speakerButton)
        private var currentMessage: ChatMessage? = null
        
        init {
            mediaImageView?.setOnClickListener {
                currentMessage?.let { message ->
                    onMediaClicked?.invoke(message)
                }
            }
            
            copyButton.setOnClickListener {
                currentMessage?.let { message ->
                    copyToClipboard(message.content)
                }
            }
            
            speakerButton.setOnClickListener {
                currentMessage?.let { message ->
                    onTtsRequested?.invoke(message)
                }
            }
        }
        
        private fun copyToClipboard(text: String) {
            val clipboard = itemView.context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            val clip = ClipData.newPlainText("Message", text)
            clipboard.setPrimaryClip(clip)
            Toast.makeText(itemView.context, "Message copied to clipboard", Toast.LENGTH_SHORT).show()
        }
        
        fun bind(message: ChatMessage) {
            currentMessage = message
            messageText.text = message.content
            
            // Handle media attachment if present
            if (message.mediaUri != null && message.mediaType == MediaType.IMAGE) {
                mediaImageView?.visibility = View.VISIBLE
                mediaImageView?.setImageURI(message.mediaUri)
            } else {
                mediaImageView?.visibility = View.GONE
            }
        }
    }

    /**
     * ViewHolder for assistant messages
     */
    class AssistantMessageViewHolder(
        itemView: View,
        private val onTtsRequested: ((ChatMessage) -> Unit)?
    ) : RecyclerView.ViewHolder(itemView) {
        
        private val messageText: TextView = itemView.findViewById(R.id.assistantMessageText)
        private val copyButton: ImageButton = itemView.findViewById(R.id.copyButton)
        private val speakerButton: ImageButton = itemView.findViewById(R.id.speakerButton)
        private var currentMessage: ChatMessage? = null
        
        init {
            copyButton.setOnClickListener {
                currentMessage?.let { message ->
                    copyToClipboard(message.content)
                }
            }
            
            speakerButton.setOnClickListener {
                currentMessage?.let { message ->
                    onTtsRequested?.invoke(message)
                }
            }
        }
        
        private fun copyToClipboard(text: String) {
            val clipboard = itemView.context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            val clip = ClipData.newPlainText("Message", text)
            clipboard.setPrimaryClip(clip)
            Toast.makeText(itemView.context, "Message copied to clipboard", Toast.LENGTH_SHORT).show()
        }
        
        fun bind(message: ChatMessage) {
            currentMessage = message
            messageText.text = message.content
            
            // Handle processing state
            if (message.isProcessing) {
                // TODO: Add a typing indicator for processing messages
            }
            
            // Handle error state
            if (message.error) {
                messageText.setTextColor(itemView.context.getColor(android.R.color.holo_red_light))
            } else {
                messageText.setTextColor(itemView.context.getColor(android.R.color.black))
            }
        }
    }

    /**
     * ViewHolder for system messages
     */
    class SystemMessageViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val messageText: TextView = itemView.findViewById(R.id.systemMessageText)
        
        fun bind(message: ChatMessage) {
            messageText.text = message.content
        }
    }

    /**
     * DiffUtil callback for efficient RecyclerView updates
     */
    class ChatMessageDiffCallback : DiffUtil.ItemCallback<ChatMessage>() {
        override fun areItemsTheSame(oldItem: ChatMessage, newItem: ChatMessage): Boolean {
            return oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: ChatMessage, newItem: ChatMessage): Boolean {
            return oldItem == newItem
        }
    }
} 